### PR TITLE
Referees 2018/automated broadcasting

### PR DESCRIPTION
**Pull request by idanstark42**
This PR is for automating the scores processing, and making the synchronization invisible to the user.
What it contains:
1. Adding dependency on ng-cookies, and saving the current page in cookies.
2. Added log writing in sendError in the utils server module
3. Small change to the file_system server module to use promises and the sendError function from the utils server module
4. Added a listening option to the ng-message module.
5. The scoring mechanism: 
 - Upon scoresheet submition, $scores.create() is called with the scoresheet data, which is sent to the server, and saved to both the scoresheet file and the scores.json file
- Every change to the scores.json file (editing, publishing/un-publishing and deleting a score) will pass through a function on the server which locks the file for that change time, and only changes the file that change, instead of completely overwriting it. This ensures concurrency.
- When a change is complete, the client receives the new scores.json and reloads it, and he sends to the rest of the clients a message via mhub to also reload the scores, keeping all the client synchronized. 
5. With the new sync mechanism, there's no need for the buttons on the top of the scorekeeping view, and they are removed.
6. An option for automatic publishing was added in the settings, in a new tab called "Advanced". When this option is enabled, when getting the approval that $scores.create worked, a new request for publishing will be sent. For that, the publishing method was passed to inside the ng-scores module. 
   NOTE: "advanced" tabs are not recommended in general in UX, might be a good idea to rethink this tab in the close future, maybe even in this PR.
7. An option for automatic broadcasting was added, and it can set in the ranking page, by pressing on one of the "Auto-broadcast" button. When selected, the stage selected will be broadcast automatically upon each edit of the scores (publishing, editing, or deleting) except for creation when there is no auto-publishing, because there's no new published scores, so there's no need for broadcasting.

This is kind of a big change, so I think we should think about it carefully. Any notes about the design or the coding itself will be happily accepted.
I tested this only on my windows computer, so I don't know how the package lockfile works on linux or mac.

### CODE REVIEW

**Comment by poelstra** [comment]
First of all, thanks for all the hard work! :+1: 

Your way of splitting the operations into CRUD operations on the server is a good solution, but it is a fundamental change in the design of the app: we deliberately kept the server 'dumb' (or at least tried to), being not more than a remote filesystem. Now, personally, I'm not against moving things to the server-side, but this does mean we're going to abandon one of the initial goals of this project: to run everything as e.g. a single iPhone app. I would really like to have discussions with @kmeesters and @rikkertkoppes before we continue into this direction.

In addition to the broadcast stuff, this PR contains a number of other things. In my opinion, the PR as it currently stands is too big/diverse to be reviewed/accepted in its current form. Therefore, please create separate PRs for the individual (largely unrelated) improvements into separate PRs, so that we can simplify and speed up the review/merge.

For example, the cookie-stuff (please also explain in that PR why it is needed/what is does?), the writeFile improvement (which is a good one), logging stuff, etc. Even the CRUD scores stuff can be seen as independent from the Refresh Scores stuff, so it would greatly help to have these separate too.

Another thing to note is that most automated tests now fail (132 out of 279 tests). You should be able to run `npm test` on your own machine to run the tests yourself. Note that the server did not have tests before (it probably should have had a few by now, would have made your and my life easier on the last PR ;)), but especially now we're going to move stuff over there, we will have to start making tests for it as well. We can definitely help in setting up the groundwork for that (shouldn't be that hard).

(I'll defer actual review/tests of the code to the individual PRs.)

Again, I'd like to stress that I think you're doing an excellent job, and I'm sure we'll get most if not all of your changes merged!

**Comment by rikkertkoppes** [comment]
I have serious doubts against moving (more) stuff to the server. If anything, I'd suggest the opposite.

Having a reliable network connection and a reliable server is not guaranteed. I think one of the core features of the critical components (which are judging and scoring) should be that they should be robust against network problems. You simply cannot pause a tournament or lose a score in a network partition.

I think we should move more to the client for these applications. That is, client side storing of results (in localstorage) and synchronizing with the server whenever there is a connection (because redundant storage is also important; having everything on a device that can be dropped or ran over is also not ideal)

This is no simple thing to implement though. If anything, an online client-server model is much easier to develop and reason about. With distributed smart clients and partitioned data storage you can run into all sorts of problems that need to be dealt with, like race conditions, merge conflicts and eventual consistency. Also reconnect and synchronization logic.

I realize this is a rather strong standpoint, especially since I am currently not in a position to contribute very much. Also, following my standpoint does probably mean slower development, more intricate code, a steeper learning curve for those eager to join the development and maybe even a sense of [clueyness](https://waitbutwhy.com/2016/05/clueyness-a-weird-kind-of-sad.html)

If I were to advice one thing than it would be to consider what would happen if the server is unreachable for, say, half an hour. How would that affect your tournament? If there are ways to handle it, we should maybe just document that failsafe and continue on the path to move stuff to the server. After all, it does make everything a lot simpler.

**Comment by idanstark42** [comment]
Hi all, thank you a lot for your quick responses! I want you to know that I take everything you say very seriously.

First of all, I will split it into several PRs, I completely agree that it's too broad and it can be simply divided.

Next thing, The tests are failing because of the big change in the PR, after splitting it I will test it separately and fix the tests in each PR.

And last and most important (and you can see I wrote the most about here), the server-client approach. 
I completely understand the idea of being able to cope with the bad network conditions, and I really love it. And if we could have the best of both world, it would be great. But I don't think it's possible, so I think we need to rethink if this is the right priority. 
We need to make the program that will be the best for our users, i.e. the tournament officials. We're used to network errors that are caused by the old program, which used up the entire bandwidth. But the program you built here is fast and easy on the network, so network failure is of very little chance.
To us, as programmers and tech people. it's very logical to have a "sync" button. But for non-technological people, it's a burden that make them ask "Why didn't they just do it automatically?".  We know that automatic sync is much more complicated then that, but they don't. And it makes it not fun to use the program. I think that users will feel better with a program that's more automatic but also more prone to network errors. Notice: not data loss, just network errors. because if the client can't send scoresheet to the server, then it wont clear the page, and and error pops up saying so.
As for falling servers, if you look at other mechanisms that use data decentralization (indexing engines, NO-SQL database, large websites), they do it with multiple servers, not by saving data in the client. Users can close the browser whenever they want, and it's not in our control. The server on the other hand, should be running at all times. If the server is down, it's an error, if the client is down it's a use case. decentralizing data to clients is, in my opinion, more risky then relaying on a single server.
(Of curse, the ideal situation is to relay on a decentralized DB with a decentralized server, but that seems a little out of reach for us)

I'm sorry if I'm coming off harsh, but I really think that this system is a great one, and this approach in the server-client matter is something that's holding us back from really hitting the full extent of the technologies that are implemented here.

**Comment by sleemans** [comment]
Hi all,

A small coin from my side: in my experience as a head referee, the problem is often the wifi: the average secondary-school Eduroam connection does not handle hundreds of devices very well. So, yes, network failure does happen, or even worse: on the moment supreme, it times out.

Browsers have the annoying habit of emptying the page when something goes wrong (or the user presses the wrong button), and in a time-constrained round, there’s usually no time to score everything again. Thus, please ( ;) ), keep the scores on until the scores have been submitted (and this has been confirmed). I have seen landing pages come in the middle and give a nice 400 code back, even though it’s the landing page answering.

Thus, I side between Rikkert and Idan: I think the clients should be smart enough to work without any server connection. That is, scores should be kept in a (inspectable!) submission queue and the client should be able to start scoring the next match without network connectivity. However, I see no objection in handling the administrative stuff (like rankings, corrections, publishing and so on) on the server (that is also where the interesting synchronisation issues might occur).

**Comment by rikkertkoppes** [comment]
> **idanstark42:** But for non-technological people, it's a burden

Agreed, I think this is one of the major hurdles that prevents further adoption

> **sleemans:** the problem is often the wifi

This is more of a problem than server failure in my opinion. As @idanstark42 said, servers can be built quite robust. It's what we do.

> **sleemans:** Browsers have the annoying habit of emptying the page when something goes wrong

This is the current situation and a result of bad (or less than ideal) software, built only around the happy path. But not necessarily the case.

What I would propose is something along these lines

- any action is applied pessimistically with regard to on-device storage, but optimistically wrt to server storage.
- actions to be sent to the server are locally stored in a queue, when a server response (ack) is received, the queue is shifted

So no sync button but automatic syncing. As actions are applied locally, scores are locally available (and inspectable if built that way), unless local storage goes wrong. In that case, there may be some serious problem with the device (storage full or something), which would prevent continuation.

Whenever there is network, anything that has not been acknowledged yet by the server is sent out again and removed from the queue when acknowledged. We may even provide a view to show pending actions.

Of course it is not as simple as this. For this to work, all actions should be idempotent (to prevent double submissions, a problem we currently have). This can be done by adding a sequence number. Also, we need to think about what to do with mhub messages that could not be sent out. Do we resend or drop them? Are these messages still relevant after 5 min of wifi outage?

File: server_modules/scores.js [hunk 1]
```diff
@@ -15,12 +18,56 @@ function reduceToMap(key) {
     }
 }
 
+function summery(scoresheet) {
```
Comment (rikkertkoppes) [inline server_modules/scores.js]: spelling: summary

File: server_modules/file_system.js [hunk 2]
```diff
@@ -51,13 +49,15 @@ exports.readFile = function(file) {
...
 
-exports.writeFile = function(file, contents, cb) {
+exports.writeFile = function(file, contents) {
     file = exports.resolve(file);
 
-    var dir = path.dirname(file);
-    mkdirp(dir, function(err) {
-        if (err) return cb(err);
-        fs.writeFile(file, contents, cb);
+    return Q.promise(function(resolve, reject) {
```
Comment (rikkertkoppes) [inline server_modules/file_system.js]: Promises are native to node and browsers these days, I think we can safely remove Q by now

File: server_modules/scores.js [hunk 3]
```diff
@@ -1,7 +1,10 @@
+var lockfile = require('lockfile');
 var utils = require('./utils');
 var fileSystem = require('./file_system');
 var Q = require('q');
 
+const scoreSummeryFile = 'scores.json';
```
Comment (rikkertkoppes) [inline server_modules/scores.js]: spelling: summary

File: server_modules/utils.js [hunk 4]
```diff
@@ -1,16 +1,19 @@
...
+exports.root = __dirname + '/../';
+
+exports.sendError = function(res) {
+    return function(err) {
+        log.error(err.message);
+        res.status(err.status).send(err.message);
+    }
+}
+
+if (!String.prototype.format) {
```
Comment (rikkertkoppes) [inline server_modules/utils.js]: with ES6 template strings, I guess we can remove String.format

**Comment by idanstark42** [comment]
Hi everyone. Thank you again for your replies: I take them very seriously.
I want to point you in the direction of the new PRs that are actually this PR split into smaller pieces:
1. [file system promises](https://github.com/FirstLegoLeague/fllscoring/pull/254)
2. [cookie pages - less important](https://github.com/FirstLegoLeague/fllscoring/pull/255)
3. [scores sync](https://github.com/FirstLegoLeague/fllscoring/pull/256)
The rest of the changes I chose to leave here, for as you see after we merged the other PRs that it's not a lot. 

After the discussion here, I went with the middle ground and added the ngStorage to the front-end of the code, and used it to save the scoresheet in an event where they cannot be sent. When a scoresheet is successfully submitted, the scoresheets waiting in the local storage will be sent as well. This all sits in the third PR: scores sync.

Again, please tell me any comments that you have on any of these. Going easy on me will not lead us anywhere, so please, even the tiniest concerns may and should be said, otherwise it's problem that can catch us unprepared down the road.

**Comment by idanstark42** [comment]
This PR is dependent of the scores_sync PR that I just moved on from, and I'll created a separate PR for each part of the sync

File: server_modules/utils.js (hunk 4 above)
Comment (poelstra) [inline server_modules/utils.js]: Yes, definitely remove overloading of the String prototype! :)

**Comment by poelstra** [comment]
Small comment on @rikkertkoppes remark:
> Of course it is not as simple as this. For this to work, all actions should be idempotent (to prevent double submissions, a problem we currently have).

This used to be the case a long time ago, but no longer is. The scoresheet submission is now always idempotent.

**Pull request closed without merging**